---
title: "Clonal analysis"
author: "Balazs Koncz"
date: '2020 11 24 '
output: html_document
---

#Setup

```{r}
setwd("d:/CloudStation/mygit/fitness-modell-bk/")
library(Rfast)
library(purrr)
library(furrr)
#library(pbapply)
ext_folder1 = "d:/CloudStation/fitness-model-ext/"
```


#Collect peptides

```{r}
proteome = read.delim(file = paste0(ext_folder1, "objects/uniprot-filtered-organism__Homo+sapiens+(Human)+[9606]_.tab"), stringsAsFactors = F)

annfiles = list.files("D:/CloudStation/fitness-model/create_neoantigens/annotated/", full.names = T)

dtann1 = read.table("D:/CloudStation/fitness-model/create_neoantigens/annotated/AL4602.txt", header = F, stringsAsFactors = F, col.names = c("Uploaded_variation","Location","Allele","Gene","Feature","Feature_type","Consequence","cDNA_position","CDS_position","Protein_position","Amino_acids","Codons","Existing_variation","Extra"))
dtann = matrix(NA, nrow = 0, ncol = ncol(dtann1), dimnames = list(NULL,colnames(dtann1)))
rm(dtann1)

for(i in annfiles) {
  tempannfile = read.table(i, header = F, stringsAsFactors = F, col.names = colnames(dtann))
  dtann = rbind(dtann, tempannfile)
}
rm(tempannfile, annfiles, i)

dtann = dtann[dtann$Consequence == "missense_variant",]
dtann$sampleid = sapply(dtann$Uploaded_variation, function(x) strsplit(x, "_")[[1]][5])
dtann = dtann[,c("sampleid", "Uploaded_variation","Feature","Protein_position","Amino_acids","Extra")]
class(dtann$Protein_position) = "integer"

# no_cores <- detectCores() - 1
# c1 <- makeCluster(no_cores)

dtann$origaa = sapply(dtann$Amino_acids, function(x) substr(x,1,1))
dtann$mutaa = sapply(dtann$Amino_acids, function(x) substr(x,3,3))
dtann$origseq = pbsapply(dtann$Extra, function(x) gsub("WildtypeProtein=", "", grep("WildtypeProtein=", strsplit(x, ";")[[1]], value = T)))
dtann$validmut = pbapply(dtann, 1, function(x) substr(x[9],as.numeric(x[4]),as.numeric(x[4])) == x[7])
dtann = subset(dtann, validmut == TRUE) #all of them
dtann$protlen = nchar(dtann$origseq)
dtann$mutseq = pbapply(dtann, 1, function(x) paste0(substr(x[9],1,as.numeric(x[4])-1),x[8],substr(x[9],as.numeric(x[4])+1,x[11])))

#determine -8 +8 region
aaregion = t(pbapply(dtann,1,function(x) {
  temppos = as.numeric(x[4])
  templen = as.numeric(x[11])
  temporigseq = x[9]
  tempmutseq = x[12]
  if(temppos<9) {
    out = c(substr(temporigseq,1,temppos+8),substr(tempmutseq,1,temppos+8))
  } else if(temppos>templen-8) {
    out = c(substr(temporigseq,temppos-8,templen),substr(tempmutseq,temppos-8,templen))
  } else {
    out = c(substr(temporigseq,temppos-8,temppos+8),substr(tempmutseq,temppos-8,temppos+8))
  }
  out
}))
colnames(aaregion) = c("origreg", "mutreg")
dtann = cbind(dtann, aaregion)
rm(aaregion)

#STATUS - reviewed or unreviewed
# clusterExport(c1, varlist = "proteome")
# status = parSapply(cl = c1, dtann$Feature, FUN = function(x) ifelse(length(grep(x,proteome$Ensembl.transcript)) > 0, proteome$Status[grepl(x,proteome$Ensembl.transcript)],NA))
# stopCluster(c1)
# rm(c1, no_cores)
# 
# status = pbsapply(X = dtann$Feature, FUN = function(x) ifelse(length(grep(x,proteome$Ensembl.transcript)) > 0, proteome$Status[grepl(x,proteome$Ensembl.transcript)],NA), simplify = T, USE.NAMES = F, cl = c1)
# 
# dtann$status = sapply(dtann$Feature, function(x) ifelse(length(grep(x,proteome$Ensembl.transcript)) > 0, proteome$Status[grepl(x,proteome$Ensembl.transcript)],NA))
# dtann = subset(dtann, status == "reviewed")

dtann = unique(dtann[,c("sampleid", "Uploaded_variation", "Feature","origreg","mutreg")])
dtann = dtann[!grepl('X|U', dtann$origreg) & !grepl('X|U', dtann$mutreg),]
dtann = dtann[nchar(dtann$origreg)>=9,]
save(dtann, file = paste0(ext_folder1, "06out/dtann_01"))
```

#Genotypes

```{r}
load("D:/CloudStation/cohort12/c1c2")
c1c2 = tidyr::separate(data = c1c2, col = "HLA.Class.I.Alleles", into = c("a1", "a2", "b1", "b2", "c1", "c2"), sep = ",", remove = F)
c1c2 = c1c2[,c("Sample", "a1", "a2", "b1", "b2", "c1", "c2")]
c1c2[,2:7] = apply(c1c2[,2:7], 2, function(x) paste0("HLA-", substr(x,1,3),":",substr(x,4,5)))
vallen = read.delim(paste0(ext_folder1, "objects/TableS2_Revised.txt"), stringsAsFactors = F)
c1c2 = c1c2[!c1c2$Sample %in% vallen$Sample,]
vallen = vallen[vallen$Sample %in% dtann$sampleid,]
c1c2 = c1c2[c1c2$Sample %in% dtann$sampleid,]
#unique(dtann$sampleid[!dtann$sampleid %in% c1c2$Sample])
cohort = rbind(c1c2, vallen)
save(cohort, file = paste0(ext_folder1, "objects/genotypes"))
rm(c1c2, vallen)
```

#Collect alleles & peptides

```{r}
load(paste0(ext_folder1, "objects/genotypes"))
ids = cohort$Sample

for(i in ids) {
  print(i)
  dir.create(path = paste0(ext_folder1, "06out/binding/", i))
  tempdtann = dtann[dtann$sampleid == i,]
  mers = c(unique(unlist(lapply(tempdtann$origreg, function(x) substring(x, 1:(nchar(x)-8), 9:nchar(x))),use.names = F)),unique(unlist(lapply(tempdtann$mutreg, function(x) substring(x, 1:(nchar(x)-8), 9:nchar(x))), use.names = F)))
  writeLines(text = mers, con = paste0(ext_folder1, "06out/binding/", i, "/nonamers.pep"), sep = "\n")
  alleles = as.character(cohort[cohort$Sample == i,2:7])
  writeLines(text = alleles, con = paste0(ext_folder1, "06out/binding/", i, "/alleles.txt"), sep = "\n")
}
rm(i, tempdtann, mers, alleles)
```

#Determine binding peptides

```{r}
load(paste0(ext_folder1, "06out/dtann_01"))
ids = unique(dtann$sampleid)
load("D:/CloudStation/fitness-model-ext/objects/pentamerfreq_tcem")
# no_cores <- detectCores() - 1
# c1 <- makeCluster(no_cores)
# clusterExport(cl = c1, varlist = c("dtann","pentamerfreq_tcem","ext_folder1","Rfast"))

plan(multisession, workers = 6)

patprop_pfreq <- future_map(.x = ids, .f = function(x) {
  patient = x
  uploaded_variation = dtann$Uploaded_variation[dtann$sampleid == x]
  feature = dtann$Feature[dtann$sampleid == x]
  load(paste0(ext_folder1, "06out/binding_matrices/binding_matrix_", x))
  #ORIG
  origreg = dtann$origreg[dtann$sampleid == x]
  origmers = lapply(origreg, function(a) substring(a, 1:(nchar(a)-8), 9:nchar(a)))
  origaffmins = sapply(origmers, function(b) {
    cm = colMins(binding_matrix$aff[,b], value = T)
    names(cm) = b
    cm
  })
  orig_affweak_n = sapply(origaffmins, function(d) sum(d<500))
  orig_affweak_medpf = sapply(origaffmins, function(e) median(pentamerfreq_tcem[substr(names(e)[e<500],4,8)]))
  orig_affstr_n = sapply(origaffmins, function(f) sum(f<50))
  orig_affstr_medpf = sapply(origaffmins, function(g) median(pentamerfreq_tcem[substr(names(g)[g<50],4,8)]))
  
  origrpmins = sapply(origmers, function(h) {
    cm = colMins(binding_matrix$rp[,h], value = T)
    names(cm) = h
    cm
  })
  orig_rpweak_n = sapply(origrpmins, function(i) sum(i<2))
  orig_rpweak_medpf = sapply(origrpmins, function(j) median(pentamerfreq_tcem[substr(names(j)[j<2],4,8)]))
  orig_rpstr_n = sapply(origrpmins, function(k) sum(k<.5))
  orig_rpstr_medpf = sapply(origrpmins, function(l) median(pentamerfreq_tcem[substr(names(l)[l<.5],4,8)]))
  
  #MUT
  mutreg = dtann$mutreg[dtann$sampleid == x]
  mutmers = lapply(mutreg, function(m) substring(m, 1:(nchar(m)-8), 9:nchar(m)))
  mutaffmins = sapply(mutmers, function(n) {
    cm = colMins(binding_matrix$aff[,n], value = T)
    names(cm) = n
    cm
  })
  mut_affweak_n = sapply(mutaffmins, function(o) sum(o<500))
  mut_affweak_medpf = sapply(mutaffmins, function(p) median(pentamerfreq_tcem[substr(names(p)[p<500],4,8)]))
  mut_affstr_n = sapply(mutaffmins, function(q) sum(q<50))
  mut_affstr_medpf = sapply(mutaffmins, function(r) median(pentamerfreq_tcem[substr(names(r)[r<50],4,8)]))
  
  mutrpmins = sapply(mutmers, function(s) {
    cm = colMins(binding_matrix$rp[,s], value = T)
    names(cm) = s
    cm
  })
  mut_rpweak_n = sapply(mutrpmins, function(t) sum(t<2))
  mut_rpweak_medpf = sapply(mutrpmins, function(u) median(pentamerfreq_tcem[substr(names(u)[u<2],4,8)]))
  mut_rpstr_n = sapply(mutrpmins, function(v) sum(v<.5))
  mut_rpstr_medpf = sapply(mutrpmins, function(w) median(pentamerfreq_tcem[substr(names(w)[w<.5],4,8)]))
  out = list(patient, uploaded_variation=uploaded_variation, feature=feature, 
             origreg=origreg, origmers=origmers, origaffmins=origaffmins, 
             orig_affweak_n=orig_affweak_n, orig_affweak_medpf=orig_affweak_medpf, 
             orig_affstr_n=orig_affstr_n, orig_affstr_medpf=orig_affstr_medpf,
             origrpmins=origrpmins, 
             orig_rpweak_n=orig_rpweak_n, orig_rpweak_medpf=orig_rpweak_medpf, 
             orig_rpstr_n=orig_rpstr_n, orig_rpstr_medpf=orig_rpstr_medpf,
             mutreg=mutreg, mutmers=mutmers, mutaffmins=mutaffmins, 
             mut_affweak_n=mut_affweak_n, mut_affweak_medpf=mut_affweak_medpf, 
             mut_affstr_n=mut_affstr_n, mut_affstr_medpf=mut_affstr_medpf,
             mutrpmins=mutrpmins, 
             mut_rpweak_n=mut_rpweak_n, mut_rpweak_medpf=mut_rpweak_medpf, 
             mut_rpstr_n=mut_rpstr_n, mut_rpstr_medpf=mut_rpstr_medpf)
  out
})

save(patprop_pfreq, file = paste0(ext_folder1, "06out/patprop_pfreq"))

patprop_pfreq[[1]][2]





rawdata = read.table(file = paste0(ext_folder1, "objects/raw_vcf/AL4602.vcf"), sep = "\t", col.names = c("CHROM","POS","ID","REF","ALT","QUAL","FILTER","INFO","FORMAT","NORMAL","TUMOR"))



#NEMOK
rawdata$orig_affweak_medtf = sapply(rawdata$ID, function(x) ifelse(grepl("_",dtann$origtcems_affweak[dtann$Uploaded_variation == x]), median(pentamerfreq_tcem[strsplit(dtann$origtcems_affweak[dtann$Uploaded_variation == x],"_")[[1]]]),median(pentamerfreq_tcem[dtann$origtcems_affweak[dtann$Uploaded_variation == x]])))

```

